### PR TITLE
Implement Access Control 

### PR DESCRIPTION
**Pull request by kylehuynh205**
## About this PR: 

- Implement configurations for the Access control with Group 
- Enable private file system and switch default file system to private file system to have all assets be protected with Group
- Switch derivatives actions to private file. (ie. https://islandora.traefik.me/admin/config/system/actions/configure/image_generate_a_service_file_from_an_original_file)
- Add [a patch](https://raw.githubusercontent.com/digitalutsc/private_files_adapter/main/scripts/openseadragon.authentication.patch) to enable OpenSeadragon View include an JWT token whenever it sends a request to Cantaloupe. (Already had [a PR](https://github.com/Islandora/openseadragon/pull/40)) 

## Required modifications from isle-dc 

- Add extra steps to create a private file system directory. 
    - In isle-dc, download the Makefile's patch file at https://raw.githubusercontent.com/digitalutsc/override_permission_file_entity/main/Makefile_accesscontrol.patch
    - Apply the patch file by `patch -p1 < Makefile_accesscontrol.patch`
- In [sample.env](https://github.com/Islandora-Devops/isle-dc/blob/development/sample.env), please change:
   - `CANTALOUPE_DELEGATE_SCRIPT_ENABLED=true`
   - `CANTALOUPE_HTTPSOURCE_LOOKUP_STRATEGY=ScriptLookupStrategy`
   ==> This can prevent 403 error in OpenSeaDragon Viewer when it views an non-public image for authorized users (even with Admin users).

### Islandora Group Module

- Module Link: https://github.com/digitalutsc/islandora_group
- Based from Danny's initial works, by tagging nodes, then it will move the node and its media into Groups, and [Group](https://www.drupal.org/project/group) will handle the access control. 
- this module will extend from Danny work to apply access control for node and media separately and provide an UI to move nodes and media to group(s)

image

image

### Group Solr

- Module Link: https://github.com/digitalutsc/group_solr
- Purpose: prevent restricted node(s) not being included in Search Results and Facet results counter. This ensure the accuracy for the Advanced Search module.
- Components:
    - Provide a Search API processor to check access control of nodes and media with Group when being indexed to Solr.
    - image
    - Provide a filter in Search API Solr based View to check the access control of node when view is loading the results in Search View: https://islandora.traefik.me/admin/structure/views/view/solr_search_content (fixed the issue with search results count in View.
    - Screenshot 2023-07-04 at 1 47 29 PM

### Private File Adapter

- Module Link: https://github.com/digitalutsc/private_files_adapter
- Implement hook_file_access to check access control in private file when Drupal  received the requests from Cantaloupe server. 

### Requirements:

To ingest node and media with access control with Islandora Workbench:

- After the build completed, visit https://islandora.traefik.me/admin/group and create a new Group. 
- In https://islandora.traefik.me/admin/structure/taxonomy/manage/islandora_access/overview, obtain the term ID 
- In the [Spreadsheet](https://docs.google.com/spreadsheets/d/1DV0Ka0ZafZq3RgCn0x_AetFtlmnMuNFltzck2QL0ApY/edit?usp=sharing), Add column called `field_access_terms`, and paste term ID of that Term associated with that newly created Group.
- In Islandora Workbench [workbench_utils.py](https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py), please add the follwing code **after** [line 3165](https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py#L3165) and [line 3191](https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py#L3191) 

````
                "field_access_terms": [{
                    "target_id": 55,
                    "target_type": 'taxonomy_term'
                }]
````
Note: Replace `55` with the tid of term associated with newly created Group

## Further development

- Currently derivatives are not automatically moved to the group(s) of their parent node.
- Restricted content's thumbnail not loading in Mirador https://github.com/digitalutsc/islandora_mirador/issues/8

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -158,6 +158,22 @@ This should get you a starter Islandora site with:
 * RDF and JSON-LD mappings for miscellaneous entities to support storage in
 Fedora, Triplestore indexing and client requests.
 
+### Private file system
+
+* In Drupal container of isle-dc, create private file system directory
```
Comment (rosiel) [inline README.md:163]: The Starter Site is not specific to ISLE-dc.

File: composer.json [hunk 2]
```diff
@@ -25,10 +25,19 @@
...
         },
-	{
-	    "type": "vcs",
-	    "url": "https://github.com/digitalutsc/advanced_search.git"
-	}
+	    {
+            "type": "package",
+            "package": {
+              "name": "islandora/advanced_search",
+              "version": "2.0.0-beta3",
```
Comment (rosiel) [inline composer.json:32]: Why specify this commit? The starter site should update with new versions (non-breaking).

File: composer.json [hunk 3]
```diff
@@ -58,13 +67,16 @@
...
+        "islandora/advanced_search": "^2.0.0@beta",
         "islandora/controlled_access_terms": "^2",
         "islandora/islandora": "^2.8.1",
         "islandora/openseadragon": "^2",
         "islandora/views_nested_details": "^1.0",
         "library/pdf.js": "^2.4",
-        "mjordan/islandora_workbench_integration": "^1.0"
+        "mjordan/islandora_workbench_integration": "^1.0",
+        "digitalutsc/group_solr": "^1.0.0-beta",
+        "digitalutsc/islandora_group": "^1.0.0-beta",
```
Comment (rosiel) [inline composer.json:78]: usually the syntax is @beta (2 lines)/

Comment (kylehuynh205) [reply composer.json]: Fixed with the commit https://github.com/kylehuynh205/islandora-starter-site/commit/abc572691cebab9571b186eaf8795fa54af29361

File: README.md (hunk 1 above)
Comment (kylehuynh205) [reply README.md:163]: Will test it with playbook and isle-site-template as well.

**Comment by kylehuynh205** [comment]
Redone with https://github.com/Islandora-Devops/islandora-starter-site/pull/100

**Pull request closed without merging**